src/cockpit/channels/dbus.py

```diff
@@ -172,11 +171,54 @@ class DBusChannel(Channel):
     matches = None
     name = None
     bus = None
+    owner = None
 
     # This needs to be a fair mutex so that outgoing messages don't
     # get re-ordered.  asyncio.Lock is fair.
     watch_processing_lock = asyncio.Lock()
 
+    def add_match(self, rule, handler):
+        def sync_handler(message):
+            task = asyncio.create_task(handler(message))
+            self.tasks.add(task)
+            task.add_done_callback(self.tasks.discard)
+        self.matches.append(self.bus.add_match(rule, sync_handler))
+
+    async def setup_name_owner_tracking(self):
+        def send_owner(owner):
+            # We must be careful not to send duplicate owner
+            # notifications. cockpit.js relies on that.
+            if self.owner != owner:
+                self.owner = owner
+                self.send_message(owner=owner)
+        async def handler(message):
+            if message.get_member() == "NameOwnerChanged":
+                name, old, new = message.get_body()
+                if name == self.name:
+                    send_owner(owner=new if new != "" else None)
+        self.add_match("type='signal',sender='org.freedesktop.DBus',path='/org/freedesktop/DBus',interface='org.freedesktop.DBus'", handler)
+        try:
+            unique_name, = await self.bus.call_method_async("org.freedesktop.DBus",
+                                                            "/org/freedesktop/DBus",
+                                                            "org.freedesktop.DBus",
+                                                            "GetNameOwner", "s", self.name)
+        except BusError as error:
+            if error.name == "org.freedesktop.DBus.Error.NameHasNoOwner":
+                # Try to start it.  If it starts successfully, we will
+                # get a NameOwnerChanged signal (which will set
+                # self.owner) before StartServiceByName returns.
+                try:
+                    await self.bus.call_method_async("org.freedesktop.DBus",
+                                                     "/org/freedesktop/DBus",
+                                                     "org.freedesktop.DBus",
+                                                     "StartServiceByName", "su", self.name, 0)
+                except BusError as error:
+                    logger.debug("Failed to start service '%s': %s", self.name, error.message)
+            else:
+                logger.debug("Failed to get owner of service '%s': %s", self.name, error.message)
+        else:
+            send_owner(unique_name)
```

Comment:
Can't we just send a `Ping` on the well-known name and check the sender field of the reply message?

Reply:
...but, do we really want to implicitly start it?  This stuff always seemed so nasty to me... :(

Reply:
Does that even have a chance to work? Normally, D-Bus autostart service files should do this automatically ("bus activation"), for services where it's intended and declared. Unless you explicitly connect with the NO_AUTOSTART flag. But some services like NetworkManager really shouldn't get auto-activated; why would a manual StartServiceByName suddenly work then, if it didn't work the first time around?

Reply:
I have copied this from GDBus name tracking, which is used by the C bridge (IIRC, it's been a while).

Reply:
We have some code that relies on this to figure out whether a service is available or not.  For example, the Storage page relies on the channel being closed with "not-found" to figure out that stratis2 isn't available and we should be using stratis3.